File: String/796. Rotate String.cpp
/*
796. Rotate String
*/

// TC:O(m + n) SC:O(m + n)
class Solution {
public:
    bool rotateString(string s, string goal) {
        int n = s.size(); 
        int m = goal.size();
        if(n != m) return false;
        s = s + s;
        // z function 先算出自己的狀況
        n = s.size();
        auto cal_z = [&](string& goal) -> vector<int> {
            vector<int> z(m);
            z[0] = m;
            for(int i = 1, c = 1, r = 1, len; i < m; i++) {
                len = r > i ? min(z[i-c], r-i) : 0;
                while(i + len < m && goal[i] == goal[i+len]) {
                    len++;
                }
                if(i + len > r) {
                    c = i;
                    r = i + len;
                }
                z[i] = len;
            }

            return z;
        };

        vector<int> z = cal_z(goal);
        
        auto cal_e = [&](string& t, string& s) -> bool {
            vector<int> e(n);
            for(int i = 0, r = 0, c = 0, len; i < n; i++) {
                len = r > i ? min(z[i-c], r-i) : 0;
                while(i + len < n && s[i+len] == t[len]) {
                    len++;
                }
                if(i + len > r) {
                    c = i;
                    r = i + len;
                }
                e[i] = len;
                if(e[i] == (int)t.size()) return true;
                if(i > n/2+1) return false;
            }

            return false;
        };

        return cal_e(goal, s);
    }
};

File: String/796. Rotate String.py
# TC:O(n) SC:O(n)
class Solution:
    def rotateString(self, s: str, goal: str) -> bool:
        def kmp_search(text:str, pattern:str) -> int:
            m = len(pattern)
            lsp = [0] * m
            for i in range(1, m):
                j = lsp[i - 1]
                while j > 0 and pattern[j] != pattern[i]:
                    j = lsp[j - 1]
                lsp[i] = j + (1 if pattern[j] == pattern[i] else 0)
            n = len(text)
            dp = [0] * n
            dp[0] = 1 if text[0] == pattern[0] else 0
            cnt = 0
            for i in range(1, n):
                j = dp[i - 1]
                if j == m:
                    j = lsp[j - 1]
                while j > 0 and pattern[j] != text[i]:
                    j = lsp[j - 1]
                dp[i] = j + (1 if pattern[j] == text[i] else 0)
                if dp[i] == len(pattern):
                    cnt += 1
            return cnt
        return len(s) == len(goal) and kmp_search(s + s, goal) > 0
